src/compiler/parser.ts

```diff
@@ -3480,7 +3518,7 @@ module ts {
                 case SyntaxKind.VoidKeyword:
                     return parseVoidExpression();
                 case SyntaxKind.LessThanToken:
-                    return parseTypeAssertion();
+                    return (sourceFile.jsxFactory && tryParse(() => parseJSXElement(true))) || parseTypeAssertion();

```

Comment:
this is going to be a big hit given that we do tihs on every type assertion.  We'll need to preprocess the file to determine what the set of possible close tag names are (i.e. search for all occurrences of "</id>").  Then, based on that table, we can determine if we need to do the costly parseJSX path or not.